File: lib/domain/entities/deck.dart
import 'package:equatable/equatable.dart';
import 'package:flash_mastery/domain/entities/flashcard_type.dart';

/// Deck entity representing a collection of flashcards under a folder.
class Deck extends Equatable {
  final String id;
  final String name;
  final String? description;
  final String? folderId;
  final int cardCount;
  final FlashcardType? _type;
  final DateTime createdAt;
  final DateTime updatedAt;

  /// Returns the deck type, defaulting to vocabulary if null
  FlashcardType get type => _type ?? FlashcardType.vocabulary;

  const Deck({
    required this.id,
    required this.name,
    this.description,
    this.folderId,
    this.cardCount = 0,
    FlashcardType? type,
    required this.createdAt,
    required this.updatedAt,
  }) : _type = type;

  Deck copyWith({
    String? id,
    String? name,
    String? description,
    String? folderId,
    int? cardCount,
    FlashcardType? type,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Deck(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      folderId: folderId ?? this.folderId,
      cardCount: cardCount ?? this.cardCount,
      type: type ?? _type,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        folderId,
        cardCount,
        _type,
        createdAt,
        updatedAt,
      ];
}

File: flash_mastery_api/src/main/java/com/flash/mastery/entity/Deck.java
package com.flash.mastery.entity;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

import com.flash.mastery.entity.enums.FlashcardType;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Deck extends BaseAuditEntity {

    private String name;
    private String description;
    private int cardCount;

    @Default
    private FlashcardType type = FlashcardType.VOCABULARY;

    private UUID folderId;

    @ToString.Exclude
    private Folder folder;

    @ToString.Exclude
    @Default
    private Set<Flashcard> flashcards = new HashSet<>();
}
